### PR TITLE
0016 frame flight protocol

### PR DESCRIPTION
**Pull request by Fazanza**
Please leave any comments about the code that u find bad. I will get to as soon as I can. Also will work on the document in the upcoming days

### CODE REVIEW

File: synnax/pkg/distribution/framer/encoder/encoder.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+		index += 16
+	}
+
+	if channelFlag == 1 {
+		returnStruct.Keys = m.keys
+	}
+
+	for i := 0; i < len(m.keys); i++ {
+		if channelFlag == 0 {
+			if index < len(src) && channel.Key(binary.LittleEndian.Uint32(src[index:])) == m.keys[i] {
```
Comment (emilbon99) [inline synnax/pkg/distribution/framer/encoder/encoder.go:147]: If statement could be better structured. Just check for the converse and use continue? (or is it inverse). Anyway, you don't need an else statement here.

File: synnax/pkg/distribution/framer/encoder/encoder_test.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+				Keys:   keySlice[0:2],
+				Series: []telem.Series{series1, series5},
+			}
+			byteArray, err := cd.Encode(testStruct)
+			returnStruct, err := cd.Decode(byteArray)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(testStruct).To(Equal(returnStruct))
+		})
+	})
+})
```
Comment (emilbon99) [inline synnax/pkg/distribution/framer/encoder/encoder_test.go:113]: I want performance tests!

**Review by emilbon99 — CHANGES_REQUESTED** [review]
Some changes to make. I'd also like to see performance profiling in terms of heap allocation, especially when compared to JSON.

File: synnax/pkg/distribution/framer/encoder/encoder.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+		curDataSize, index               int             = -1, 0
+		startTime, endTime               telem.TimeStamp = 0, 0
+		returnArray                      []byte
+		sizeFlag, alignFlag, channelFlag int = 1, 1, 0
+	)
+
+	channelFlag = map[bool]int{false: 0, true: 1}[len(m.keys) == len(src.Keys)]
+
+	for i := 0; i < len(m.keys); i++ {
+		if index >= len(src.Keys) || src.Keys[index] != m.keys[i] {
```
[the remark this answers is not in the document]
Comment (Fazanza) [reply synnax/pkg/distribution/framer/encoder/encoder.go:47]: The way this is set up, the index will go over the len of the src keys, and was causing an error. Had to implement this condition to stop that.

**Comment by emilbon99** [comment]
LGTM

**Pull request merged**